File: FCBARCA.py
# O(1)
mod = int(1e9+7)

for _ in range(int(input())):
	n, k = map(int, input().split())
	ans = (k ** n + k * ((-1) ** n)) // (k + 1)
	print(ans % mod)
  
  
  


# O(n)
mod = int(1e9+7)

for _ in range(int(input())):
	n, k = map(int, input().split())
	ls = [0] * (n + 1)
	ls[0], ls[1] = 1, 0
	for i in range(2, n+1):
		ls[i] = ((k-1) * ls[i-1] + k * ls[i-2]) % mod
	print(ls[n])

File: FCBARCA.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'
long long mod = 1e9+7;

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
#endif

	int t;
	cin >> t;
	while (t--) {
		int n, k;
		cin >> n >> k;
		long long g[n+1];
		g[0] = 1, g[1] = 0;
		for (int i = 2; i <= n; i++)
			g[i] = ((k-1) * g[i-1] + k * g[i-2]) % mod;
		cout << g[n] << endl;
	}
}
